localstack-core/localstack/services/sns/v2/provider.py

```diff
@@ -118,6 +164,360 @@ def set_topic_attributes(
+        principal = DUMMY_SUBSCRIPTION_PRINCIPAL.format(
+            partition=get_partition(context.region), account_id=context.account_id
+        )
+        subscription_arn = create_subscription_arn(topic_arn)
+        subscription = SnsSubscription(
+            # http://docs.aws.amazon.com/cli/latest/reference/sns/get-subscription-attributes.html
+            TopicArn=topic_arn,
+            Endpoint=endpoint,
+            Protocol=protocol,
+            SubscriptionArn=subscription_arn,
+            PendingConfirmation="true",
+            Owner=context.account_id,
+            RawMessageDelivery="false",  # default value, will be overridden if set
+            FilterPolicyScope="MessageAttributes",  # default value, will be overridden if set
+            SubscriptionPrincipal=principal,  # dummy value, could be fetched with a call to STS?
+        )
+        if sub_attributes:
+            subscription.update(sub_attributes)
+            if "FilterPolicy" in sub_attributes:
+                filter_policy = (
+                    json.loads(sub_attributes["FilterPolicy"])
+                    if sub_attributes["FilterPolicy"]
+                    else None
+                )
+                if filter_policy:
+                    validator = FilterPolicyValidator(
+                        scope=subscription.get("FilterPolicyScope", "MessageAttributes"),
+                        is_subscribe_call=True,
+                    )
+                    validator.validate_filter_policy(filter_policy)
+
+                store.subscription_filter_policy[subscription_arn] = filter_policy
+
+        store.subscriptions[subscription_arn] = subscription
+
+        topic_subscriptions = store.topic_subscriptions.setdefault(topic_arn, [])
+        topic_subscriptions.append(subscription_arn)
+
+        # store the token and subscription arn
+        # TODO: the token is a 288 hex char string
+        subscription_token = encode_subscription_token_with_region(region=context.region)
+        store.subscription_tokens[subscription_token] = subscription_arn
+
+        response_subscription_arn = subscription_arn
+        # Send out confirmation message for HTTP(S), fix for https://github.com/localstack/localstack/issues/881
+        if protocol in ["http", "https"]:
+            message_ctx = SnsMessage(
+                type=SnsMessageType.SubscriptionConfirmation,
+                token=subscription_token,
+                message=f"You have chosen to subscribe to the topic {topic_arn}.\nTo confirm the subscription, visit the SubscribeURL included in this message.",
+            )
+            publish_ctx = SnsPublishContext(
+                message=message_ctx,
+                store=store,
+                request_headers=context.request.headers,
+                # TODO: add topic attributes once they are ported from moto to LocalStack
+                # topic_attributes=vars(self._get_topic(topic_arn, context)),
```

Comment:
nit: I believe those should now be present in the topic, right?

Reply:
yes, but as I was not yet sure about the PublishContext and the exact effects of the code, I would rather do this together with publish

Reply:
makes sense 👍 I believe it will be enough to pass the `topic["attributes"]`, let's check it later! 👍